// rc-banner-anim/lib/Arrow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Arrow = function (_Component) {
  (0, _inherits3['default'])(Arrow, _Component);

  function Arrow() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Arrow);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = Arrow.__proto__ || Object.getPrototypeOf(Arrow)).call.apply(_ref, [this].concat(args))), _this), _this.onClick = function (e) {
      if (e.stopPropagation) {
        e.stopPropagation();
      }
      _this.props[_this.props.arrowType](e);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Arrow, [{
    key: 'render',
    value: function render() {
      var className = this.props.className;
      var defaultClass = className + '-default';
      className = (className + ' ' + (this.props.prefixCls || '')).trim();
      className = !this.props['default'] ? className : (className + ' ' + defaultClass).trim();
      className = className + ' ' + this.props.arrowType;
      var props = (0, _extends3['default'])({}, this.props, this.props.componentProps);
      ['arrowType', 'next', 'prev', 'elemHeight', 'component', 'componentProps', 'default', 'prefixCls'].forEach(function (key) {
        return delete props[key];
      });
      props.className = className;
      props.onClick = this.onClick;
      props.style = props.style || {};
      props.style.top = this.props.elemHeight / 2 + 'px';
      return _react2['default'].createElement(this.props.component, props, this.props.children);
    }
  }]);
  return Arrow;
}(_react.Component);

Arrow.propTypes = {
  children: _propTypes2['default'].any,
  style: _propTypes2['default'].object,
  className: _propTypes2['default'].string,
  prefixCls: _propTypes2['default'].string,
  component: _propTypes2['default'].any,
  arrowType: _propTypes2['default'].string,
  'default': _propTypes2['default'].bool,
  next: _propTypes2['default'].func,
  prev: _propTypes2['default'].func,
  elemHeight: _propTypes2['default'].number,
  componentProps: _propTypes2['default'].object
};
Arrow.defaultProps = {
  component: 'div',
  className: 'banner-anim-arrow',
  componentProps: {}
};

Arrow.isBannerAnimArrow = true;

exports['default'] = Arrow;
module.exports = exports['default'];
