// rc-banner-anim/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.switchChildren = exports.setAnimCompToTagComp = exports.animType = exports.Thumb = exports.Element = exports.Arrow = undefined;

var _BannerAnim = require('./BannerAnim');

var _BannerAnim2 = _interopRequireDefault(_BannerAnim);

var _Arrow = require('./Arrow');

var _Arrow2 = _interopRequireDefault(_Arrow);

var _Element = require('./Element');

var _Element2 = _interopRequireDefault(_Element);

var _Thumb = require('./Thumb');

var _Thumb2 = _interopRequireDefault(_Thumb);

var _anim = require('./anim');

var _anim2 = _interopRequireDefault(_anim);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

// export this package's api
_BannerAnim2['default'].Arrow = _Arrow2['default'];
_BannerAnim2['default'].Element = _Element2['default'];
_BannerAnim2['default'].Thumb = _Thumb2['default'];
_BannerAnim2['default'].animType = _anim2['default'];
_BannerAnim2['default'].setAnimCompToTagComp = _utils.setAnimCompToTagComp;
_BannerAnim2['default'].switchChildren = _utils.switchChildren;

exports['default'] = _BannerAnim2['default'];
exports.Arrow = _Arrow2['default'];
exports.Element = _Element2['default'];
exports.Thumb = _Thumb2['default'];
exports.animType = _anim2['default'];
exports.setAnimCompToTagComp = _utils.setAnimCompToTagComp;
exports.switchChildren = _utils.switchChildren;
