// rc-banner-anim/lib/Thumb.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Thumb = function (_Component) {
  (0, _inherits3['default'])(Thumb, _Component);

  function Thumb() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Thumb);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = Thumb.__proto__ || Object.getPrototypeOf(Thumb)).call.apply(_ref, [this].concat(args))), _this), _this.getDefaultThumb = function () {
      var children = [];
      for (var i = 0; i < _this.props.length; i++) {
        children.push(_react2['default'].createElement('span', { key: i }));
      }
      return children;
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Thumb, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var className = 'banner-anim-thumb';
      var defaultClass = className + '-default';
      className = (className + ' ' + (this.props.prefixCls || '')).trim();
      className = !this.props['default'] ? className : (className + ' ' + defaultClass).trim();
      var children = this.props['default'] ? this.getDefaultThumb() : this.props.children;
      if (this.props.length && (0, _utils.toArrayChildren)(children).length !== this.props.length) {
        console.warn('The thumbnail length and the images length different.'); // eslint-disable-line
      }
      var childToRender = (0, _utils.toArrayChildren)(children).map(function (item, i) {
        var props = (0, _extends3['default'])({}, item.props);
        props.onClick = function (e) {
          if (e.stopPropagation) {
            e.stopPropagation();
          }
          _this2.props.thumbClick(i);
        };
        props.className = ((props.className || '') + ' ' + (_this2.props.active === i ? 'active' : '')).trim();
        return _react2['default'].cloneElement(item, props);
      });
      var props = (0, _extends3['default'])({}, this.props, this.props.componentProps);
      ['length', 'thumbClick', 'active', 'default', 'component', 'componentProps', 'prefixCls'].forEach(function (key) {
        return delete props[key];
      });
      props.className = className;
      return _react2['default'].createElement(this.props.component, props, childToRender);
    }
  }]);
  return Thumb;
}(_react.Component);

Thumb.propTypes = {
  children: _propTypes2['default'].any,
  style: _propTypes2['default'].object,
  prefixCls: _propTypes2['default'].string,
  component: _propTypes2['default'].any,
  thumbClick: _propTypes2['default'].func,
  'default': _propTypes2['default'].bool,
  length: _propTypes2['default'].number,
  active: _propTypes2['default'].number,
  componentProps: _propTypes2['default'].object
};
Thumb.defaultProps = {
  component: 'div',
  componentProps: {},
  thumbClick: function thumbClick() {}
};
Thumb.isBannerAnimThumb = true;

exports['default'] = Thumb;
module.exports = exports['default'];
