// rc-texty/lib/TextyAnim.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _rcTweenOne = require('rc-tween-one');

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _animTypes = require('./animTypes');

var _animTypes2 = _interopRequireDefault(_animTypes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var __rest = undefined && undefined.__rest || function (s, e) {
    var t = {};
    for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
    }if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
        if (e.indexOf(p[i]) < 0) t[p[i]] = s[p[i]];
    }return t;
};

function transformArguments(arg, e) {
    var result = void 0;
    if (typeof arg === 'function') {
        result = arg(e);
    } else {
        result = arg;
    }
    return result;
}

var TextyAnim = function (_React$Component) {
    (0, _inherits3['default'])(TextyAnim, _React$Component);

    function TextyAnim() {
        (0, _classCallCheck3['default'])(this, TextyAnim);

        var _this = (0, _possibleConstructorReturn3['default'])(this, _React$Component.apply(this, arguments));

        _this.getChildrenToRender = function (str) {
            if (!str) {
                return [];
            }
            var split = _this.props.split;

            var t = split ? split(str) : str.split(''); // Array(str.length).fill(1);
            return t.map(function (c, i) {
                return React.createElement("span", { key: c + '-' + i.toString() }, c);
            });
        };
        _this.getEnterOrLeave = function (e, genre, length) {
            var _this$props = _this.props,
                mode = _this$props.mode,
                type = _this$props.type,
                enter = _this$props.enter,
                appear = _this$props.appear,
                interval = _this$props.interval,
                duration = _this$props.duration;

            if (!appear && genre === 'enter' || length < 0) {
                return null;
            }
            var cb = (0, _extends3['default'])({}, e, { type: genre });
            var $duration = transformArguments(duration, cb);
            var $interval = transformArguments(interval, cb);
            var delay = void 0;
            if (typeof interval === 'function') {
                // function 下 mode 无效；
                delay = $interval;
            } else {
                switch (mode) {
                    case 'reverse':
                        delay = (length - e.index) * $interval;
                        break;
                    case 'sync':
                        delay = 0;
                        break;
                    case 'random':
                        delay = length * $interval * Math.random();
                        break;
                    default:
                        delay = e.index * $interval;
                }
            }
            delay += transformArguments(_this.props.delay, cb);
            var genreType = genre === 'enter' ? 'from' : 'to';
            var custom = _this.props[genre] || enter;
            if (custom && typeof custom === 'function') {
                custom = custom((0, _extends3['default'])({}, e, { delay: delay }));
            } else if (!custom) {
                custom = _animTypes2['default'][type];
            }
            if (custom.enter) {
                custom = custom[genre] || custom.etner;
            }
            if (Array.isArray(custom)) {
                return custom.map(function (item, i) {
                    if (!i && (!item.duration || item.type === 'set')) {
                        return item;
                    } else if (i === 1) {
                        var preItem = custom[0];
                        if (!preItem.duration || item.type === 'set') {
                            return (0, _extends3['default'])({ delay: delay }, item);
                        }
                    }
                    return (0, _extends3['default'])({ delay: !i ? delay : 0 }, item);
                });
            }
            return (0, _extends3['default'])({ delay: delay, duration: $duration, type: genreType }, custom);
        };
        return _this;
    }

    TextyAnim.prototype.render = function render() {
        var _classnames,
            _this2 = this;

        var _a = this.props,
            prefixCls = _a.prefixCls,
            type = _a.type,
            className = _a.className,
            enter = _a.enter,
            mode = _a.mode,
            duration = _a.duration,
            delay = _a.delay,
            interval = _a.interval,
            children = _a.children,
            split = _a.split,
            props = __rest(_a, ["prefixCls", "type", "className", "enter", "mode", "duration", "delay", "interval", "children", "split"]);
        var getChildrenToRender = this.getChildrenToRender(children);
        var classNames = (0, _classnames3['default'])(prefixCls, (_classnames = {}, _classnames[type] = type && !enter, _classnames[className] = !!className, _classnames));
        return React.createElement(_rcTweenOne.TweenOneGroup, (0, _extends3['default'])({}, props, { className: classNames, enter: function enter(e) {
                return _this2.getEnterOrLeave(e, 'enter', getChildrenToRender.length - 1);
            }, leave: function leave(e) {
                return (
                    // 出场时 children 是没有， 需取 group 里的  keysToLeave
                    _this2.getEnterOrLeave(e, 'leave', _this2.tweenGrooup.keysToLeave.length - 1)
                );
            }, ref: function ref(c) {
                _this2.tweenGrooup = c;
            } }), getChildrenToRender);
    };

    return TextyAnim;
}(React.Component);

exports['default'] = TextyAnim;

TextyAnim.defaultProps = {
    type: 'top',
    mode: 'smooth',
    prefixCls: 'texty',
    component: 'div',
    delay: 0,
    interval: 50,
    appear: true
};
module.exports = exports['default'];
